### PR TITLE
Package Ubuntu Touch click package

### PR DESCRIPTION
**Pull request by jonnius**
Solves #294.

Whats working:
- Building click package (see README, needs a [clickable patch](https://gitlab.com/clickable/clickable/merge_requests/128))
- Everything seems to work so far.

Whats missing:
- Does not work well with dark system theme (white text on white background)
- Some graphical glitches (some artifacts on icons, Text in dropdown selection boxes is too big)
- Optional dependency `mimic1` is missing

### CODE REVIEW

File: packaging/ubports/manifest.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
+    "name": "pure-maps.jonius",
+    "description": "Navigation app",
```
Comment (rinigus) [inline packaging/ubports/manifest.json]: This we can replace with "Maps and navigation"

**Comment by rinigus** [comment]
Thank you very much, looking forward to see this!

**Comment by jonnius** [comment]
Would you be fine with adding the dependencies as gitmodules, so we can [fully automate](https://blog.bhdouglass.com/clickable/tutorial/2019/03/19/publishing-apps-to-the-0penstore-with-gitlab-ci.html) the whole click build and publish process?

**Comment by rinigus** [comment]
Sorry, no, can't do that. For Debian that would be unacceptable, for Sailfish also very odd. So, let's keep dependencies as they are - dependency, not submodule. I have added some submodules, but its mainly for relatively exotic and small things.

Note that on Sailfish, we have already two programs using Mapbox QML module and lots of them use Nemo DBUS. So, maybe these will also move later to the common libraries, as other libs are.

**Comment by jonnius** [comment]
The `pure-maps` startup script contains an absolute path to the `share` folder where `pure-maps.qml` is located. This can be fixed by making this path relative. @rinigus how can I achieve that? Atm I am doing a `make PREFIX=$INSTALL_DIR INCLUDE_GPXPY=yes install`

**Comment by jonnius** [comment]
On the phone (with pure-maps script fixed) I get:
```
Loading module: 'libubuntu_application_api_touch_mirclient.so.3.0.0'
Current locale: de_DE
Loaded translation
Got library name:  "/usr/lib/arm-linux-gnueabihf/qt5/qml/io/thp/pyotherside/libpyothersideplugin.so"
QQmlApplicationEngine failed to load component
file:///opt/click.ubuntu.com/pure-maps.jonius/0.1.0/share/pure-maps/qml/pure-maps.qml:26 Type InitPage unavailable
file:///opt/click.ubuntu.com/pure-maps.jonius/0.1.0/share/pure-maps/qml/InitPage.qml:23 Type PageEmptyPL unavailable
file:///opt/click.ubuntu.com/pure-maps.jonius/0.1.0/share/pure-maps/qml/platform/PageEmptyPL.qml:27 Cannot assign to non-existent property "globalToolBarStyle"

Error loading /opt/click.ubuntu.com/pure-maps.jonius/0.1.0/share/pure-maps/qml/pure-maps.qml
```
Possibly some newer Kirigami feature?

**Comment by rinigus** [comment]
Hi, do you get plain grey screen without map? Any output on terminal? I have seen something similar on Debian. If it's the same, we can make tmp version with qt only and later switch to latest kirigami.

**Comment by rinigus** [comment]
Sorry, was reading on phone. Let me fix that issue. I will be able to do it later tonight. If you can, comment out line 27 of pageemptypl.qml

**Comment by jonnius** [comment]
After manually commenting out that line, I got
```
Current locale: de_DE
Loaded translation
Got library name:  "/usr/lib/arm-linux-gnueabihf/qt5/qml/io/thp/pyotherside/libpyothersideplugin.so"
QQmlApplicationEngine failed to load component
file:///opt/click.ubuntu.com/pure-maps.jonius/0.1.0/share/pure-maps/qml/pure-maps.qml:71 Type PositionSource unavailable
file:///opt/click.ubuntu.com/pure-maps.jonius/0.1.0/share/pure-maps/qml/PositionSource.qml:25 Type PositionSourceMapMatched unavailable
file:///opt/click.ubuntu.com/pure-maps.jonius/0.1.0/share/pure-maps/qml/PositionSourceMapMatched.qml:31 module "Nemo.DBus" is not installed

Error loading /opt/click.ubuntu.com/pure-maps.jonius/0.1.0/share/pure-maps/qml/pure-maps.qml
```
Seems like it does not find its dependencies...

**Comment by rinigus** [comment]
exactly, dependencies are not found. you would have to specify path to qml modules using `path` option of qmlrunner. we would probably have to patch pure-maps script for ubports to include it.

**Comment by rinigus** [comment]
one tip: for nemo.dbus, path should point to the folder, where nemo/dbus will be found as a relative path

**Comment by jonnius** [comment]
The issues above are solved. 

Now I get an initialization animation before the screen turns white. There are still some QML errors, see the log file: [error.log](https://github.com/rinigus/pure-maps/files/3511765/error.log)

**Comment by rinigus** [comment]
looks to be unhappy with old kirigami. maybe we should bundle latest kirigami as well. kirigami depends only on cmake extra modules and then should compile well. 

extra permissions needed probably for audio access. it also need permissions to write cache file and config files. for some reason it cannot access fallback icons.

**Comment by jonnius** [comment]
The fallback directory is not being installed together with the other icons.

From which repository should I clone Kirigami?

**Comment by rinigus** [comment]
hmm, fallback icons should be installed into icons, next to qml dir. the path that I am using is wrong for ubports, have to fix this bug. sorry for it, I'll fix it tonight

kirigami clone command is given at https://api.kde.org/frameworks/kirigami/html/index.html. lets use v5.61.0 tag. same tag for cmake with clone command at https://api.kde.org/frameworks/extra-cmake-modules/html/index.html

**Comment by jonnius** [comment]
I just tried to compile kirigami 5.61.0, but it requires Qt 5.11, while UT only ships Qt 5.9. I think shipping Qt too is not an option ;)

**Comment by rinigus** [comment]
I'll try to,see why it asks for Qt 5.11. Please attach compile log part where it asks for qt 5.11

**Comment by jonnius** [comment]
Its the CMake configuration. Don't you think it requires Qt5.11 for a reason?

```
CMake Error at CMakeLists.txt:34 (find_package):
  Could not find a configuration file for package "Qt5" that is compatible
  with requested version "5.11.0".

  The following configuration files were considered but not accepted:

    /usr/lib/arm-linux-gnueabihf/cmake/Qt5/Qt5Config.cmake, version: 5.9.5
```

**Comment by rinigus** [comment]
was it during compilation of kirigami or cmake modules?

**Comment by jonnius** [comment]
Kirigami

**Comment by rinigus** [comment]
looks like right now we can use kirigami 5.54. I have asked kirigami devs why required version was bumped. which kirigami ships with ubports?

**Comment by jonnius** [comment]
Don't know where to look that up. But judging from what people in the dev groups say its Kirigami 2.4.

**Comment by rinigus** [comment]
I will look tonight and make ubports version which is qt 5.9 only. it should be quite simple and we can later enhance it futher either via kirigami or ubports specific widgets. unless we get something positive from kirigami devs

**Comment by rinigus** [comment]
Kirigami 2.4 is an import statement, which is not exactly the version of it.

**Comment by jonnius** [comment]
On the device I found `libKF5Kirigami2.so.5.52.0`

**Comment by rinigus** [comment]
thanks, looks as old as in debian.

**Comment by rinigus** [comment]
I have changed UBPorts platform to plain QtControls (commit 6360dad6ffb1651296a22c2a41d3013d58e07182). Now all should work, as long as permissions and dependencies are OK. We don't need kirigami for now...

**Comment by jonnius** [comment]
Great! I still get an issue though: 
```
QQmlApplicationEngine failed to load component
file:///opt/click.ubuntu.com/pure-maps.jonius/0.1.0/share/pure-maps/qml/pure-maps.qml:26 Type InitPage unavailable
file:///opt/click.ubuntu.com/pure-maps.jonius/0.1.0/share/pure-maps/qml/InitPage.qml:23 Type PageEmptyPL unavailable
file:///opt/click.ubuntu.com/pure-maps.jonius/0.1.0/share/pure-maps/qml/platform/PageEmptyPL.qml:36 Type HeaderBarImpl unavailable
file:///opt/click.ubuntu.com/pure-maps.jonius/0.1.0/share/pure-maps/qml/platform/HeaderBarImpl.qml:46 Cannot assign to non-existent property "icon"
```

**Comment by rinigus** [comment]
Please try the latest Pure Maps master. 

icon property of Button was introduced from Qt 5.10 and I don't know why I was able to access it on my system with the import statements corresponding to Qt 5.9. Don't have access to the system with Qt 5.9, hence these errors. But we will iron them out eventually.

**Comment by jonnius** [comment]
Yey, it starts!

[error.log](https://github.com/rinigus/pure-maps/files/3512847/error.log)

screenshot20190818_105441453
screenshot20190818_105526106
screenshot20190818_105533876
screenshot20190818_105548162

**Comment by rinigus** [comment]
There are several errors we will have to go through.

* [ERROR] Failed to create directory for the cache database /home/phablet/.cache/pure-maps/pure-maps/mapboxgl-qml-cache.db [directory that was attempted to create: /home/phablet/.cache/pure-maps/pure-maps]

You would have to allow to make files and folders for Pure Maps

* There are some DBus permissions that need to be sorted, but we can do it later

* [ ERROR ]  "{QSGRenderThread}[Setup]: loading style failed: HTTP status code 401"

This is why you don't see map. Please register at Mapbox.com as a developer and make your own API key. That key has to be used either for packaging (see instructions at https://github.com/rinigus/pure-maps#packaging) or for private use (specify in Preferences)

* There are some other errors that I will have to address later

**Comment by rinigus** [comment]
PS: Buttons on map look big, these we will have to reduce as well. later :)

**Comment by jonnius** [comment]
In [these directories](https://docs.ubports.com/en/latest/appdev/guides/writeable-dirs.html) we have write access. Don't you use Qt methods to find the cache directory? Path should be `~/.cache/pure-maps.jonius` in our case.

**Comment by rinigus** [comment]
Well, it makes at `/home/phablet/.cache/pure-maps` - which Pure Maps expects to be the full name. not with extension ".jonius"

**Comment by jonnius** [comment]
On Ubuntu Touch, the full name is always `<appname>.<maintainer>`. I think I can't avoid that.

**Comment by rinigus** [comment]
OK. I have updated Makefile for better support of alternative names. Please use `make NAME=pure-maps.jonius ...." while you install it. There is still #298 that will have to be fixed to keep all config settings remembered, but I will deal with it later tonight

**Comment by jonnius** [comment]
Thanks for that. But actually the app name should still be `pure-maps`, especially the desktop file should be `pure-maps.desktop`. There is a difference in UT between an app name used inside the app (`pure-maps`) and the full app name (`pure-maps.jonius`) used to identify the app and its paths.

**Comment by jonnius** [comment]
Anyway, with manually fixed desktop file name and a Mapbox API key added to `apikeys.py` I got another error message about a missing property `name`: [error.log](https://github.com/rinigus/pure-maps/files/3513375/error.log) and still can't see the map.

**Comment by rinigus** [comment]
Yes, I will have to fix name and other missing properties. I'll be able to do so after 1-2 hours. 

From errors it looks like connection is still rejected - maybe API key is wrong (HTTP status code 401). 

In respect of app name: desktop, exe, icons, all stay as "pure-maps". Only dirs are based on "full name", right?

**Comment by jonnius** [comment]
About app name: that is correct.

About the API key: I have a file `tool/apikeys.py` with this content:
```
Keys = {
    # https://developer.foursquare.com [old key]
    #"<FOURSQUARE_CLIENT>": "REPLACE WITH FOURSQUARE_CLIENT KEY",
    #"<FOURSQUARE_SECRET>": "REPLACE WITH FOURSQUARE_SECRET KEY",

    # mapbox.com
    "<MAPBOX_KEY>": "blablabla",
    
    # http://open.mapquestapi.com
    #"<MAPQUEST_KEY>": "REPLACE WITH MAPQUEST KEY",

    # https://geocoder.opencagedata.com/api [old key]
    #"<OPENCAGE_KEY>": "REPLACE WITH OPENCAGE KEY",

    # https://docs.stadiamaps.com/
    #"<STADIAMAPS_KEY>": "REPLACE WITH STADIAMAPS KEY",

    # here.com
    #"<HERE_APP_ID>": "REPLACE WITH HERE APP ID",
    #"<HERE_APP_CODE>": "REPLACE WITH HERE APP CODE",
}
```
Is that correct? Should I use the _Default public token_ from Mapbox.com or create a new one? And if so how should I configure that one? E.g. its asking for URLs to be allowed...

**Comment by rinigus** [comment]
Re API keys: yes, that's correct. Just be sure that full key is copied into it. Also check, if you can see this key in `poor/keystore.py` in the installed files

I will look into the other errors, so the error message is shown properly and you could choose other map provider, such as OpenTopoMap

**Comment by rinigus** [comment]
Re Default public token: yes, you could use that for now, until you plan to release it. Then we can discuss options to keep you safe

**Comment by rinigus** [comment]
Looks like you have to run

```
tools/manage-keys inject .
```

before `make install`. After that, run

```
tools/manage-keys strip .
```

**Comment by rinigus** [comment]
You can now specify FULLNAME and NAME of application separately. In my understanding, this should work

```
make FULLNAME=pure-maps.jonius NAME=pure-maps other_options install 
```

as expected in UBPorts. Now will try to fix other issues

**Comment by jonnius** [comment]
Would you just look at this? :smile: 
screenshot20190818_220301935

**Comment by rinigus** [comment]
Looking good! I am still going through error log. Also icons are huge for some reason in UBPorts...

**Comment by rinigus** [comment]
I have fixed bugs with starting geocoder and routing pages. Please try to run again and attach error log. That way I can see if there are more errors

**Comment by jonnius** [comment]
Many more things seem to work now. I just clicked on most buttons and menus to test. Scaling is really weird, as some elements are huge while others are tiny. [error.log](https://github.com/rinigus/pure-maps/files/3516006/error.log)

**Comment by rinigus** [comment]
For dbus, I opened #299

You are missing Stadia maps api key. Register there and add as you did for Mapbox. This will enable routing.

I also suggest to register on Foursquare. That will give access to nearby search.

In general, you would probably need to register for all covered services. With Sailfish and Linux (flathub) users, the only problematic one is Mapbox in terms of getting over quota. I asked users to get their private key for that service. I do provide a key in the package, but if it gets close to quota, I just disable it and make a new one for a next release. That way active users tend to register and occasional ones can still use most of the time.

From log, looks like you tested with some raster tiles as well. if its Here, api key is needed.

Offline maps will work only when we get osm scout server ported. Should be simpler than pure maps :)

So, from logs, no more errors.

As for scaling, all parameters should be in qml/platform.ubports/StylerPL.qml . This should allow us to fix scaling too. Maybe we should use ubports theme module for it. Let me read about it

**Comment by rinigus** [comment]
I am setting up clickable environment on my PC and maybe you could tell me how to make Pure Maps run using `clickable desktop`.

**Comment by jonnius** [comment]
Just for the protocol: See the README I have added.

Adding `-scale` to qmlrunner fixes the scaling in the menus, but makes scaling in the map view worse. 

Independent of the scaling issues there are some artifacts in the menu navigation icons:
screenshot20190819_203525506

**Comment by rinigus** [comment]
Let's fix it without scale parameter. I will look into your screenshots and see whats wrong

**Comment by rinigus** [comment]
I have set the sizes using ubuntu components. text size is still set ad hock via units object, but should be reasonable. now everywhere where we see something off, I will have to look into.

latest master has new styler. please test without `-scale` parameter.  I hope that map buttons are ok now.

**Comment by jonnius** [comment]
Map buttons look good now :+1: 

screenshot20190819_220709318

**Comment by jonnius** [comment]
But menu is broken: [error.log](https://github.com/rinigus/pure-maps/files/3517499/error.log)

screenshot20190819_220800253

**Comment by jonnius** [comment]
Wait, the menu is there and at a reasonable size. It is just light gray on white background :laughing:

**Comment by jonnius** [comment]
This just happens when setting system them to dark. Seems like the text is respecting that by becoming white, while the background is not.

**Comment by rinigus** [comment]
Yes, it is respecting the color palette from Ubuntu Components. I am not setting background for controls, expect that the theme will do that.

Maybe we should merge it as it is and later continue on the master branch? That would make it easier for me to test/develop

**Comment by jonnius** [comment]
Its up to you. Feel free to merge :)

**Comment by rinigus** [comment]
Thank you very much, merged

**Pull request merged**